include/deal.II/base/vectorization.h

```diff
@@ -1278,57 +1279,53 @@ class VectorizedArray<float>
                               VectorizedArray<float> *out)
 {
+  // Similar to the double case, we perform the work on smaller entities. In
+  // this case, we start from 128 bit arrays and insert them into a full 512
+  // bit index. This reduces the code size and register pressure because we do
+  // shuffles on 4 numbers rather than 16.
   const unsigned int n_chunks = n_entries / 4;
-  for (unsigned int outer = 0; outer < 16; outer += 8)
+  for (unsigned int i = 0; i < n_chunks; ++i)
     {
-      for (unsigned int i = 0; i < n_chunks; ++i)
-        {
-          __m128 u0 = _mm_loadu_ps(in + 4 * i + offsets[0 + outer]);
-          __m128 u1 = _mm_loadu_ps(in + 4 * i + offsets[1 + outer]);
-          __m128 u2 = _mm_loadu_ps(in + 4 * i + offsets[2 + outer]);
-          __m128 u3 = _mm_loadu_ps(in + 4 * i + offsets[3 + outer]);
-          __m128 u4 = _mm_loadu_ps(in + 4 * i + offsets[4 + outer]);
-          __m128 u5 = _mm_loadu_ps(in + 4 * i + offsets[5 + outer]);
-          __m128 u6 = _mm_loadu_ps(in + 4 * i + offsets[6 + outer]);
-          __m128 u7 = _mm_loadu_ps(in + 4 * i + offsets[7 + outer]);
-          // To avoid warnings about uninitialized variables, need to initialize
-          // one variable with zero before using it.
-          __m256 t0, t1, t2, t3 = _mm256_set1_ps(0.F);
-          t0        = _mm256_insertf128_ps(t3, u0, 0);
-          t0        = _mm256_insertf128_ps(t0, u4, 1);
-          t1        = _mm256_insertf128_ps(t3, u1, 0);
-          t1        = _mm256_insertf128_ps(t1, u5, 1);
-          t2        = _mm256_insertf128_ps(t3, u2, 0);
-          t2        = _mm256_insertf128_ps(t2, u6, 1);
-          t3        = _mm256_insertf128_ps(t3, u3, 0);
-          t3        = _mm256_insertf128_ps(t3, u7, 1);
-          __m256 v0 = _mm256_shuffle_ps(t0, t1, 0x44);
-          __m256 v1 = _mm256_shuffle_ps(t0, t1, 0xee);
-          __m256 v2 = _mm256_shuffle_ps(t2, t3, 0x44);
-          __m256 v3 = _mm256_shuffle_ps(t2, t3, 0xee);
-          *reinterpret_cast<__m256 *>(
-            reinterpret_cast<float *>(&out[4 * i + 0].data) + outer) =
-            _mm256_shuffle_ps(v0, v2, 0x88);
-          *reinterpret_cast<__m256 *>(
-            reinterpret_cast<float *>(&out[4 * i + 1].data) + outer) =
-            _mm256_shuffle_ps(v0, v2, 0xdd);
-          *reinterpret_cast<__m256 *>(
-            reinterpret_cast<float *>(&out[4 * i + 2].data) + outer) =
-            _mm256_shuffle_ps(v1, v3, 0x88);
-          *reinterpret_cast<__m256 *>(
-            reinterpret_cast<float *>(&out[4 * i + 3].data) + outer) =
-            _mm256_shuffle_ps(v1, v3, 0xdd);
-        }
-      for (unsigned int i = 4 * n_chunks; i < n_entries; ++i)
-        for (unsigned int v = 0; v < 8; ++v)
-          out[i][v + outer] = in[offsets[v + outer] + i];
+      // To avoid warnings about uninitialized variables, need to initialize
+      // one variable with some old value.
+      __m512 t0, t1, t2, t3 = out[0].data;
```

Comment:
Does it make sense to read a value here instead of initializing to ``{}`` or something?

Reply:
Out of the 250 lines I changed you spot the one that is critical. :smile: 

I would have liked to use `t3 = {}` like I did for the other parts but it turns out that gcc-9 generates wrong code (at least to the extent that `vectorization_05.release` fails because it wrongly assumes alignment of the `in + offsets[v]` points to 16 byte which cannot be guaranteed. It works with all other compilers I have tried on godbolt. I am in the process of filing a compiler bug report once I have the case reduced and verified that I'm not doing something obviously wrong first.

In any case, I will add a short remark here.

Reply:
Alright. Compiler bugs are scary.

Can you move the declaration of these temporaries outside the loop? Or does that break performance?

Reply:
That is a good idea I should have had myself during the last 3 hours of debugging. It indeed works and it will not alter performance notably, judging from the assembly. In any case, it will be better than the spurious load. I will push it.

Reply:
It turned out that it is not enough - if the compiler is able to see that `n_entries` is 4 and thus there is no loop, it might inline the whole thing and again emit the wrong instruction. I've added the load back (now outside the loop) and linked to the bugzilla of gcc that I opened this morning so we keep track of it.